nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py

```diff
@@ -0,0 +1,665 @@
+        else:
+            init_consumed_samples = 0
+        self.init_consumed_samples = init_consumed_samples
+
+        if stage == 'predict':
+            return
+
+        # If the user wants to manually override train and validation dataloaders before calling `.fit()`
+        if self._train_dl is not None and self._validation_dl is not None:
+            return
+        self.build_train_valid_test_datasets(stage=stage)
+        if hasattr(self, '_train_ds'):
+            self.setup_training_dataloader()
+        if hasattr(self, '_validation_ds'):
+            self._validation_dl = self.setup_eval_dataloader(self._validation_ds, self.cfg.data.validation_ds)
+        if hasattr(self.cfg.data, 'test_ds'):
+            self._test_dl = self.setup_eval_dataloader(self._test_ds, self.cfg.data.test_ds)
+
+        # when using pipeline model parallel the final stage need to initialize word embeddings
+        if parallel_state.get_pipeline_model_parallel_world_size() > 1:
+            if isinstance(self.model, list):
+                for i, module in enumerate(self.model):
+                    parallel_state.set_virtual_pipeline_model_parallel_rank(i)
+                    module.sync_initial_word_embeddings()
+                parallel_state.set_virtual_pipeline_model_parallel_rank(0)
+            else:
+                self.model.sync_initial_word_embeddings()
+
+        if self.cfg.get('transformer_engine', False):
+            self.setup_transformer_engine_tp_groups()
+
+    def _build_dataset(self, data_cfg, is_train=True):
+        datasets = []
+        # Determine if we are using a single dataset or a list of datasets.
+        is_list_config = isinstance(data_cfg.file_names, ListConfig)
+        if not is_list_config:
+            raise ValueError(f"SFT train/validation datasets must be provided as a list of individual JSONL files.")
+
+        if is_train:
+            # Construct the data prefix list for `get_datasets_weights_and_num_samples()`
+            # that is of the format [weight1,file_name1,weight2,file_name2,...]
+            if data_cfg.concat_sampling_probabilities is None or not isinstance(
+                data_cfg.concat_sampling_probabilities, ListConfig
+            ):
+                raise ValueError(
+                    (
+                        f"concat_sampling_probabilities must be a ListConfig with the same number of files in file_names."
+                        f"Found: {data_cfg.concat_sampling_probabilities}"
+                    )
+                )
+
+            if len(data_cfg.get('concat_sampling_probabilities', None)) != len(data_cfg.file_names):
+                raise ValueError(
+                    (
+                        f"concat_sampling_probabilities must be of the same size as file_names.",
+                        f"Provided size {len(data_cfg.concat_sampling_probabilities)}, number of datasets {len(data_cfg.file_names)}",
+                    )
+                )
```

Comment:
Maybe can move this to megatron base model? I think gpt pretraining could use the same checks?

Reply:
GPT Pre-training data is provided differently where sampling probs and datasets are provided in a single argument. I wanted to separate it out, so it wouldn't translate directly.

Reply:
Yeah I think we should go the other way and add this to pretraining while keep backward compat for pretraining. Basically like this way of configuring it more.